### DISCUSSION TITLE
Upgrade from shlink 3.7.4 to 4.1.1 fails

### DISCUSSION THREAD

**Discussion by nellen** [shlinkio/discussions#147] [Help wanted] [answered]
### Shlink version

4.1.1

### PHP version

8.3.6

### How do you serve Shlink

Other (explain in summary)

### Database engine

MySQL

### Database version

10.6-MariaDB

### Summary

Hi there

I use multiple shlink v3 instances on a shared hosting (Hostpoint Business Webhosting, Nginx (Proxy) with Apache/PHP, Switzerland) since more than a year without a problem. Instances are protected through free Cloudflare instances.
Each instance is a separate website, with his own files and a dedicated MariaDB.

I tried to update my youngest shlink instance (installed with 3.7.3 in early 2024, updated to 3.7.4 afterwards) to 4.1.1
That's my second or third try to update after tries with shlink 4.0.1 or 4.0.2

Before that I downgrade again (backups available), I prefer to go forwared and try to fix this.

Update works like, CLI is working BUT web access through app.shlink.io does not work.

What I tried so far and did not help:
- Generate a new API key
- Switch webserver config from "Nginx (Proxy) with Apache/PHP" to "Apache/PHP*
- Switch webserver config back to "Nginx (Proxy) with Apache/PHP"
- Remove Nginx cache
- Switch Cloudflare to "development mode" (=no caching)
- Use app.shlink.io in a new private window
- Delete /data/cache/app_config.php

My generated_config.php:
```<?php
return array(
    'DB_DRIVER' => 'mysql',
    'DB_NAME' => 'redacted',
    'DB_HOST' => 'redacted',
    'DB_PORT' => '3306',
    'DB_USER' => 'redacted',
    'DB_PASSWORD' => 'redacted',
    'DB_UNIX_SOCKET' => null,
    'DEFAULT_DOMAIN' => 'redacted.link',
    'IS_HTTPS_ENABLED' => true,
    'DEFAULT_SHORT_CODES_LENGTH' => 5,
    'AUTO_RESOLVE_TITLES' => true,
    'REDIRECT_APPEND_EXTRA_PATH' => true,
    'MULTI_SEGMENT_SLUGS_ENABLED' => false,
    'SHORT_URL_TRAILING_SLASH' => true,
    'SHORT_URL_MODE' => 'strict',
    'GEOLITE_LICENSE_KEY' => 'redacted',
    'REDIRECT_STATUS_CODE' => 302,
    'DISABLE_TRACKING' => false,
    'TRACK_ORPHAN_VISITS' => true,
    'DISABLE_TRACK_PARAM' => null,
    'DISABLE_TRACKING_FROM' => null,
    'DISABLE_IP_TRACKING' => false,
    'ANONYMIZE_REMOTE_ADDR' => true,
    'DISABLE_UA_TRACKING' => false,
    'DISABLE_REFERRER_TRACKING' => false,
    'DEFAULT_BASE_URL_REDIRECT' => 'https://redacted.com',
    'DEFAULT_INVALID_SHORT_URL_REDIRECT' => null,
    'DEFAULT_REGULAR_404_REDIRECT' => null,
    'DEFAULT_QR_CODE_SIZE' => 800,
    'DEFAULT_QR_CODE_MARGIN' => 0,
    'DEFAULT_QR_CODE_FORMAT' => 'png',
    'DEFAULT_QR_CODE_ERROR_CORRECTION' => 'h',
    'DEFAULT_QR_CODE_ROUND_BLOCK_SIZE' => true,
    'DELETE_SHORT_URL_THRESHOLD' => 3,
    'BASE_PATH' => '',
    'TIMEZONE' => 'GMT+1',
    'REDIS_SERVERS' => null,
    'RUNTIME' => 'regular',
    'CACHE_NAMESPACE' => 'Shlink',
    'MATOMO_ENABLED' => false,
    'QR_CODE_FOR_DISABLED_SHORT_URLS' => false,
    'MEMORY_LIMIT' => '512M',
    'DEFAULT_QR_CODE_COLOR' => '#000000',
    'DEFAULT_QR_CODE_BG_COLOR' => '#FFFFFF',
    'DEFAULT_QR_CODE_LOGO_URL' => null,
);
```

Thanks for your help!

Regards
Sebastian

**Comment by acelaya** [comment]
Can you elaborate on what you mean by "web access through app.shlink.io does not work"?

Have you checked nginx, Apache and Shlink logs?

Have you checked the browser console for errors? And the network tab?

Do short URLs work?

**Reply by nellen** [reply]
Thanks for your questions, which I answered below.
I guess the Error in shlink-log (see below) is the problem, but I haven't found a solution for it.

> Can you elaborate on what you mean by "web access through app.shlink.io does not work"?

I get the error "Something went wrong while loading short URLs :("
<img width="1727" alt="image" src="https://github.com/shlinkio/shlink/assets/3300600/b6375603-412a-4ff0-8728-32e18a7a2d7c">


> Have you checked nginx, Apache and Shlink logs?`

webserver logs:
```
IP_F - - [15/Jun/2024:11:26:16 +0200] "GET /rest/v3/mercure-info HTTP/1.1" 500 143 "https://app.shlink.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
IP_F - - [15/Jun/2024:11:26:16 +0200] "GET /rest/v3/tags/stats HTTP/1.1" 500 143 "https://app.shlink.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
IP_F - - [15/Jun/2024:11:26:16 +0200] "GET /rest/v3/short-urls?itemsPerPage=5&orderBy=dateCreated-DESC HTTP/1.1" 500 143 "https://app.shlink.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
IP_F - - [15/Jun/2024:11:26:16 +0200] "GET /rest/v3/visits HTTP/1.1" 500 143 "https://app.shlink.io/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
```


Full **shlink logs** attached, Error below
[shlink-logs.txt](https://github.com/user-attachments/files/15846343/shlink-logs.txt)

```
[2024-06-15T11:26:16.953300+02:00]` [9f851942-0ad5-44d5-9e44-c16ab7aa260e] Shlink.ERROR - ErrorException: Attempt to read property "columnName" on array in /shlinkPATH/vendor/doctrine/orm/src/Mapping/DefaultQuoteStrategy.php:28
Stack trace:
```


> Have you checked the browser console for errors? And the network tab?`

Browser console = error 500 (internal server error)
<img width="1731" alt="image" src="https://github.com/shlinkio/shlink/assets/3300600/5ba9113b-9ffe-4568-9f8c-56eaea1d3556">

Network tab = error accessing files (I guess as expected when having an error 500)
<img width="1730" alt="image" src="https://github.com/shlinkio/shlink/assets/3300600/b36c3705-a4be-4669-b0a5-4bdcb38c4f34">


> Do short URLs work?

Unfortunately, no

**Reply by acelaya** [reply]
> Full **shlink logs** attached, Error below [shlink-logs.txt](https://github.com/user-attachments/files/15846343/shlink-logs.txt)
> 
> ```
> [2024-06-15T11:26:16.953300+02:00]` [9f851942-0ad5-44d5-9e44-c16ab7aa260e] Shlink.ERROR - ErrorException: Attempt to read property "columnName" on array in /shlinkPATH/vendor/doctrine/orm/src/Mapping/DefaultQuoteStrategy.php:28
> Stack trace:
> ```

This is the key point.

Shlink 4 updates the database ORM to the next major version, and there are probably important differences.

I suspect you brought all ORM generated stuff from the old version to the new one, which is causing this error. Try deleting everything inside `data/proxies` and see if it helps.

It could also be related with cache, but when using Nginx/Apache, restarting the server should have been enough to clear it. If you are using redis though, you'll need to clear Shlink cache (let me know if this is the case, and I'll explain you how).

In any case, shlink-installer should have taken care of that. Did you follow the update process as described here https://shlink.io/documentation/update-your-instance/#updating-a-self-hosted-instance?

**Reply by acelaya** [reply, marked as answer]
Actually, I just remembered about something.

When not using redis, Shlink uses an in-memory cache storage which is shared by all Shlink instances running in the same server. This can be a problem, so Shlink 3.7 introduced a way to define a unique cache prefix per instance, via `CACHE_NAMESPACE` config option or env var (in your config the value is "Shlink").

What's happening here is probably that your newest Shlink 4 instance is "picking-up" class metadata entries from the cache that were generated by Shlink 3.

Try setting a unique value there, and that should solve the problem.

See https://github.com/shlinkio/shlink/discussions/1862 for context.

**Reply by nellen** [reply]
Thanks a lot, @acelaya 

Indeed, I just had to change config param to something new
`'CACHE_NAMESPACE' => 'somethingNew',`

Then I deleted the _data/cache/app_config.php_ and my instance is back online.

**Reply by nellen** [reply]
> In any case, shlink-installer should have taken care of that. Did you follow the update process as described here https://shlink.io/documentation/update-your-instance/#updating-a-self-hosted-instance?

My update process is based on the process described in the link above and looks basically like this

```
mysqldump -h HOST -u "USER" -p DATABASE > DATABASE_$(date +'%Y%m%d%H%M').sql

wget https://github.com/shlinkio/shlink/releases/download/v4.1.1/shlink4.1.1_php8.3_dist.zip && unzip shlink*_dist.zip && mv myShlinkInstance myShlinkInstance.bak_$(date +'%Y%m%d') && mv ./shlink*_dist ./myShlinkInstance

cd myShlinkInstance && /usr/local/php83/bin/php vendor/bin/shlink-installer update

>> yes
>> ../myShlinkInstance_AddDateHere   (e.g. ../stag.art24.link.bak_20240130)
>> [ENTER]
>> [ENTER]
>> [ENTER]

touch .htaccess
nano .htaccess

>>> .htaccess adding <<<
## Added for Hostpoint by SNe ##
DirectoryIndex index.php index.html

AddHandler application/x-httpd-php83 .php
## ## ## ## ## ##

cd .. && rm -rf shlink*_dist.zip && rm -rf shlink*_dist

```
